architecture/2025-03-upgrade-feature-spec.md

```diff
@@ -0,0 +1,221 @@
+# Topic: Radius Control Plane Upgrades
+
+- **Author**: Yetkin Timocin (@ytimocin)
+
+## Topic Summary
+
+The Radius in-place upgrade feature aims to streamline the process of upgrading the control plane components of Radius, ensuring minimal downtime and disruption.
+
+**Radius Components (as of Mar 2025):**
+
+- Universal Control Plane
+- Radius Deployment Engine
+- Applications Resource Provider
+- Dynamic Resource Provider
+- Dashboard
+- Controller
+
+**Dependencies:**
+
+- Dapr
+- Contour
+- Postgres
+
+**Key Features:**
+
+- **Compatibility:** Ensures compatibility with new versions.
+- **Data Safety:** Takes snapshots before the upgrade to safeguard user data, with automatic rollback in case of failure.
+- **Seamless Experience:** Provides a smooth and reliable upgrade process with minimal downtime.
+
+This feature will significantly improve the user experience by automating the upgrade process, reducing the risk of errors, and maintaining system stability.
+
+### Top level goals
+
+- Deliver a seamless and intuitive upgrade experience for users.
+- Ensure a smooth and reliable upgrade process for the components and dependencies of Radius.
+- Minimize downtime and disruption during upgrades, ensuring that the Radius CLI is available as soon as possible.
+- Ensure user data safety during the upgrade process by taking snapshots and providing rollback capabilities.
+- Provide clear documentation and guidance for users performing upgrades.
+- Maintain system performance during the upgrade process to avoid significant impact on running applications.
+
+### Non-goals (out of scope)
+
+- Downgrades are not supported. The upgrade process is designed to move forward to newer versions only.
```

Comment:
The rollbacks mentioned above is effectevely going to be a downgrade back to the old version though?

Reply:
We have two states: current version and desired version.

Rollbacks will only occur if there is a failure during the `rad upgrade kubernetes` process and they are going to roll back to the current version. So, it is not going to go back to an older version.

It just means: "You wanted to upgrade to version 0.44 from 0.25 but there is a failure during that process, we are taking you back to 0.25"